base_suspend_security/models/base.py

```diff
@@ -17,3 +17,12 @@ def suspend_security(self):
                 self.env.cr,
                 BaseSuspendSecurityUid(self.env.uid),
                 self.env.context))
+
+    def sudo(self, user=SUPERUSER_ID):
+        if isinstance(self.env.uid, BaseSuspendSecurityUid):
```

Comment:
please add a comment here that this is necessary because of https://github.com/OCA/OCB/blob/12.0/odoo/api.py#L858 which is used in https://github.com/OCA/OCB/blob/12.0/odoo/models.py#L4772

Reply:
actually, this shouldn't be a problem, as we override __int__: https://github.com/OCA/server-backend/blob/8687422d0eed95a8e0cd2de1b39c621ab1e68c8b/base_suspend_security/base_suspend_security.py#L8 - can you debug where exactly the wrapper is lost?

Reply:
who has the wrapper is the self.uid, when doing sudo the `user` parameter not necessarily has the wrapper but still is important to maintain it and avoid not to break the thread.

Reply:
who has the wrapper is the self.uid, when doing sudo the `user` parameter not necessarily has the wrapper but still is important to maintain it and avoid to break the thread.
Also: 
```
>>> isinstance(int(BaseSuspendSecurityUid(1)), BaseSuspendSecurityUid)
False
```

In python 3.6:  DeprecationWarning: __int__ returned non-int (type BaseSuspendSecurityUid).  The ability to return an instance of a strict subclass of int is deprecated, and may be removed in a future version of Python.

Reply:
ah, now I see the problem. We will get this issue in other places too, as some of the trickery of this module relies on being able to survive int(). But those we can handle as they come.

Now looking closer at this, I don't believe this is the correct patch. Something like

```python
env['some.model'].suspend_security().sudo(42)
```

must run as user 42 without suspended security. With your code, we'll never be able to unsuspend security.

Is there some other way to achieve what you want?

Reply:
shouldn't we see the marker in `user` if this comes from a `suspend_security`'d search?

Reply:
> Now looking closer at this, I don't believe this is the correct patch. Something like
> 
> ```python
> env['some.model'].suspend_security().sudo(42)
> ```
> 
> must run as user 42 without suspended security. With your code, we'll never be able to unsuspend security.
> 
> Is there some other way to achieve what you want?

`BaseModel._search` does the following:

```python
        self.sudo(access_rights_uid or self._uid).check_access_rights('read')
```
If the sequence `suspend_security().sudo()` must reactivate the security, maybe the change is to test the argument to `sudo`:

```python
if isinstance(user, BaseSuspendSecurityUid):
   ...
```

Reply:
> ```python
> if isinstance(user, BaseSuspendSecurityUid):
> ```

I agree, that should work.